addons/sale/models/sale.py

```diff
@@ -24,19 +24,25 @@ def _amount_all(self):
         Compute the total amounts of the SO.
         """
         for order in self:
-            amount_untaxed = amount_tax = 0.0
+            round_curr = order.pricelist_id.currency_id.round
+            tax_grouped = {}  # we want the same behavior as invoices
+            amount_untaxed = 0.0
             for line in order.order_line:
                 amount_untaxed += line.price_subtotal
                 # FORWARDPORT UP TO 10.0
-                if order.company_id.tax_calculation_rounding_method == 'round_globally':
-                    price = line.price_unit * (1 - (line.discount or 0.0) / 100.0)
-                    taxes = line.tax_id.compute_all(price, line.order_id.currency_id, line.product_uom_qty, product=line.product_id, partner=order.partner_shipping_id)
-                    amount_tax += sum(t.get('amount', 0.0) for t in taxes.get('taxes', []))
-                else:
-                    amount_tax += line.price_tax
+                price = line.price_unit * (1 - (line.discount or 0.0) / 100.0)
+                taxes = line.tax_id.compute_all(price, line.order_id.currency_id, line.product_uom_qty, product=line.product_id, partner=order.partner_shipping_id)
+                for tax in taxes.get('taxes'):
+                    tax_amount = order.company_id.tax_calculation_rounding_method == 'round_globally' and tax['amount'] or round_curr(tax['amount'])
```

Comment:
This is really really annoying to put here.  Set_taxes is full of nice context overrides and ability to control tax rounding how you want, here it is awful/non existent.  e.g some people need different rounding dependent on whether it is an inclusive or exclusive tax.  But tbh I don't really see the point anyway, as it is just doing what calling compute all should do in the first place in compute_taxes.

But if it is broken, surely the correct place to fix tax rounding is in a function that sets/computes taxes, not a function that computes sale totals.

Reply:
In fact looking at this, it seems the only real bug is not calling compute_all for 'round_per_line' (in a default install) as compute_all does round the tax amounts to currency in that case, so there should be no need to rerround and you can trust the function return is what was intended for the tax amount.

Reply:
@gdgellatly 
In my fix there is something redundant indeed: the if round_method == "round_globally". As you rightfully pointed out compute_all should already have (or not if round is global) rounded those lines.

However, compute_all returns for each (order/invoice) line the list of taxes applied to that line. We indeed need a grouping function downstream (in invoice: directly in compute amount total) to aggregate each tax amounts

But then again, this fix will never be merged (not as is at least) since it is too big of a change in the one hand, but too little of a real fix on the other.

Reply:
@kebeclibre I get what you are doing, yes its just that line that concerned me.  Besides I think the fix is actually relatively big.

Sale Total and Fully Invoiced Total should be invariant *ceteris paribus*.  Be interesting to see a PR refusing that failing test case.